### ISSUE TITLE
aws_ec2: random AZ assignment is always the same?

### ISSUE THREAD

**Issue by pacohope** [aws/aws-cdk#22476]
### Describe the bug

When I read [the documentation on Python aws_ec2.instance](https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_ec2/Instance.html) it says:
> availability_zone (Optional[str]) – In which AZ to place the instance within the VPC. Default: - Random zone.

I am creating a VPC with 3 public subnets, 3 NAT gateways, and 3 private subnets wired up with routes to the 3 NAT gateways. But the EC2 instances deployed to that VPC are always assigned to the 'a' Availability Zone.

### Expected Behavior

When I run `cdk synth` and inspect the resulting json template, I always see the following:
`"AvailabilityZone": "us-east-2a"`

If the AZ selected is actually random, I expect it to see "us-east-2a" sometimes, "us-east-2b" sometimes, and "us-east-2c" sometimes.

### Current Behavior

Every CloudFormation output for every instance always selects the first availability zone. Do I misunderstand what "Random Zone" means?

### Reproduction Steps

This app will reproduce what I am seeing. 

1. Run `cdk synth` on this app
2. Run `grep Availability cdk.out/basisi*json` 
3. Observe that it's all `"AvailabilityZone": "us-east-2a",`

```
import os
import aws_cdk as cdk
import aws_cdk.aws_ec2 as ec2
import aws_cdk.aws_iam as iam
from constructs import Construct

class VpnInstancesStack(cdk.Stack):

    def __init__(self, vpcStack: Construct, construct_id:
        str, thing: str, env: cdk.Environment,  **kwargs) -> None:
        super().__init__(vpcStack, construct_id, env=env, **kwargs)

        thisVpc  = vpcStack.vpc
        ec2instance = ec2.Instance(self, f"i-{thing}",
            vpc=vpcStack.vpc,
            instance_type=ec2.InstanceType("m5.large"),
            machine_image=ec2.AmazonLinuxImage(),
            block_devices=[ec2.BlockDevice(
                device_name="/dev/sda1",
                volume=ec2.BlockDeviceVolume.ebs(50),
                )
            ],
            security_group = vpcStack.vpnSG,
        )

class VpcBasisStack(cdk.Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        self.vpc = ec2.Vpc(self, 'out-vpc',
            cidr = "172.17.0.0/18",
            max_azs = 3,
            subnet_configuration=[
                ec2.SubnetConfiguration(
                    name = 'pub',
                    subnet_type = ec2.SubnetType.PUBLIC,
                    cidr_mask = 26
                ),
                ec2.SubnetConfiguration(
                    name = 'pri',
                    subnet_type = ec2.SubnetType.PRIVATE_WITH_NAT,
                    cidr_mask = 20
                )
            ],
            nat_gateways = 3
        )

        # Create standard Security Group for all EC2 instances
        self.vpnSG = ec2.SecurityGroup(self, 'vpn-sg', vpc=self.vpc,
            allow_all_outbound=True, security_group_name='vpn-sg' );

app = cdk.App()
cdkEnv = cdk.Environment(account=os.getenv('CDK_DEFAULT_ACCOUNT'),
            region="us-east-2")

vpcStack = VpcBasisStack(app, "basis", env=cdkEnv )

deployList = [ "a", "b", "c" ]

for thing in deployList:
    VpnInstancesStack(vpcStack, f"i-{thing}-s", thing=thing, env=cdkEnv )

app.synth()
```

### CDK CLI Version

2.43.0 (build 487870a)

### Node.js Version

v14.18.3

### OS

macOS

### Language

Python

### Language Version

Python 3.9.14

**Comment by pacohope** [comment]
One other interesting dilemma in all this. At one point I experimented with doing something like the following in the `ec2.instance` creation:
```
availability_zone = random.choice(myVPC.availability_zones())
```

The problem with that, I think, is that it makes my CDK stack non-idempotent. Each time i run my python code, it will generate a random AZ assignment for my instance. If it picks a different AZ this time than it picked last time, it will cause CloudFormation to replace the instance (I think).

**Comment by corymhall** [comment]
Yeah I can see how "Random" could be confusing. When you first reference VPC resources in your CDK
application, the CDK will lookup the availability zones (via ec2.describeAvailabilityZones) and
cache the values in cdk.context.json. I'm assuming that `describeAvailabilityZones` will always
return the zones in the same order, but I can't find any documentation stating that.

I would be happy to accept a PR updating to documentation to something that is less confusing.

This issue has been classified as *p2*. That means a workaround is available or it is deemed a nice-to-have feature. Given the amount of work there is to do and the relative priority of this issue, the CDK team is unlikely to address it. That does not mean the issue will never be fixed! If someone from the community submits a PR to fix this issue, and the PR is small and straightforward enough, and meets the quality bars to be reviewed and merged with little effort we will accept that PR. PRs that do not build or need complex or multiple rounds of reviews are unlikely to be merged and will be closed to keep our backlog manageable.

In the mean time, remember that you can always use the escape hatch (https://docs.aws.amazon.com/cdk/v2/guide/cfn_layer.html) mechanism to have fine control over the CloudFormation output you want. We will keep the issue open for discoverability, to collect upvotes, and to facilitate discussion around this topic.

We use +1s *on this issue* to help prioritize our work, and are happy to re-evaluate the prioritization of this issue based on community feedback. You can reach out to the cdk.dev community on Slack (https://cdk.dev/) to solicit support for reprioritization.

**Comment by pacohope** [comment]
Thanks for the solid analysis. The more I think about this, the more I think the right thing is a PR for the documentation. 

In my opinion, non-determinism is a bad thing when synthesising CloudFromation. If we were to change the instance availability zone assignment to match the docs and somehow be "random", we'd have to worry about 2 things:
1. Making sure the same stack deploys the same way every time
2. People who deployed with a prior version of CDK that always did the first AZ, could suddenly have things targeting multiple AZs. That could cause termination/replacement, which would be unexpected and perhaps bad.

So instead, I'll try to write documentation that is clearer. We have to express the fact that it always chooses the first one off the list, but we also have to express the fact that nothing about the system makes the ordering of that list predictable or guaranteed to be a certain way.
